### ISSUE TITLE
Compiler hangs with DoublyLinkedList and concepts

### ISSUE THREAD

**Issue by michael72** [nim-lang/Nim#7412]
This is a somewhat constructed example, but with all the necessary ingredients (and additional output).
Strangely it all works, when not using either `DoublyLinkedList` or `iterType` - see code below:

```nim
import macros, lists, typetraits

type
  Iterable[T] = concept a
    for it in a:
      type(it) is T

# try to determine the type that is returned by the Iterable
proc iterType[T](items: Iterable[T]): T = 
  discard
    
# create a DoublyLinkedList with template argument depending on the output type of a handler proc
proc init[T, U](a: DoublyLinkedList[T], handler: proc(it: T): U): DoublyLinkedList[U] =
  initDoublyLinkedList[U]()

# create a seq with template argument depending on the output type of a handler proc
proc init[T, U](a: Iterable[T], handler: proc(it: T): U): seq[U] =
  @[]
  
# echo at compile time
macro myhint(a: untyped): untyped =
  quote:
    static:
      echo($`a`)

when isMainModule:    
  # I try to generate a sequence from another sequence depending.
  # The handler will convert the items of arr to the items of the output sequence (so the intention).
  # But the compiler crashes after initializing the list output in 'compileMe' section.
  proc handler(it: int): string = $it
  let arr = [1,2,3]

  # it works with sequence:
  var testSeq: seq[iterType(arr).type]
  var testSeqInit = init(testSeq, handler)
  myhint("************************************")
  myhint("The inner type of arr is: '" & $iterType(arr).type & "'")
  myhint("and testSeqInit is '" & $testSeqInit.type & "'")
  discard testSeqInit
  
  # using the explicit type instead of iterType... will also work:
  var testListWorks: DoublyLinkedList[int]  # not using iterType(arr).type here!
  var testListWorksInit = init(testListWorks, handler)
  myhint("and testListWorksInit is '" & $testListWorksInit.type & "'")
  discard testListWorksInit

  const compileMe = true # set to true -> compiler hangs afterwards
  when compileMe:
    var testList: DoublyLinkedList[iterType(arr).type]
    var testListInit = init(testList, handler)
    myhint("and testListInit also is '" & $testListInit.type & "' - still")
    myhint("we get here - but the compiler crashes (waits endlessly) afterwards...")
    discard testListInit  
  myhint("************************************")
  ```
The compiler hangs after printing:
```console
Hint: used config file 'C:\Bla\nim.cfg' [Conf]
Hint: system [Processing]
Hint: checkerror [Processing]
Hint: macros [Processing]
Hint: lists [Processing]
Hint: typetraits [Processing]
************************************
The inner type of arr is: 'int'
and testSeqInit is 'seq[string]'
and testListWorksInit is 'DoublyLinkedList[system.string]'
and testListInit also is 'DoublyLinkedList[system.string]' - still
we get here - but the compiler crashes afterwards...
************************************
CC: zero_functional_checkerror
CC: stdlib_lists
```

Used version is current devel - 0.18.1:
```console
C:\...>nim --version
Nim Compiler Version 0.18.1 [Windows: amd64]
Copyright (c) 2006-2018 by Andreas Rumpf

git hash: 1d9343080de8e19835c3f6568630ba759afbb94f
active boot switches: -d:release
```

**Comment by cooldome** [comment]
I also noticed that compilation can hang sometimes on windows, I don't think it is concept related.
My bet what is going: Nim have generated C code and executed C compiler, but generated C code does not compile. The error report by C compiler is too large (exceeds certain threshold), Nim waits forever for an exit code of C compiler, but C compiler execution paused until error output is consumed.

**Comment by michael72** [comment]
@cooldome 
Do you think this is a windows problem?
I have cleaned the nimcache and got a proper C file from the code above. The gcc call that is printed with verbosity settings works, but the nim compilation seems to be stuck with the same gcc call.

**Comment by michael72** [comment]
OK - the compiler hang is fixed with [#7472](https://github.com/nim-lang/Nim/pull/7472)
will post another probably simplified issue - because nim still does not complain, but the C-compiler does
